### PR TITLE
chore(testing): add test discovery mode for pytest

### PR DESCRIPTION
**Pull request by gnufede**
## What

Adds a test discovery mode to the ddtrace/testing pytest plugin. When `DD_TEST_OPTIMIZATION_DISCOVERY_ENABLED=true`, pytest collects all tests and writes them as JSON Lines to a file, then exits without running any test. This is used by [ddtest](https://github.com/DataDog/ddtest) during its plan phase to enumerate the test suite before deciding which tests to run or skip.

## Why

ddtest needs a language-agnostic way to enumerate tests for a given project. For Ruby, `datadog-ci-rb` provides this via a discovery mode. This PR adds the equivalent for pytest, producing the same JSONL format so ddtest's planner can parse both without special-casing the language.

## Configuration

| Env var | Default | Description |
|---|---|---|
| `DD_TEST_OPTIMIZATION_DISCOVERY_ENABLED` | `false` | Enable discovery mode |
| `DD_TEST_OPTIMIZATION_DISCOVERY_FILE` | `.testoptimization/tests-discovery/tests.json` | Output file path |

## Output format

Each line is a JSON object with:
- `name` — test name
- `suite` — test suite (file path, dot-separated)
- `module` — module (directory, dot-separated)
- `parameters` — JSON-encoded parameters for parametrised tests, or `null`
- `suiteSourceFile` — relative path to the source file

Tests decorated with `pytest.mark.skip` or `pytest.mark.skipif` with a truthy non-string condition are excluded from the output.

## Other changes

- Moves `_get_test_parameters_json` and `_encode_test_parameter` from `plugin.py` to `utils.py` to avoid a circular import with `_discovery.py`
- Registers `DD_TEST_OPTIMIZATION_DISCOVERY_ENABLED` and `DD_TEST_OPTIMIZATION_DISCOVERY_FILE` in the ddtrace settings registry
- CI Visibility initialisation is suppressed when discovery mode is active (no spans emitted)

### CODE REVIEW

File: ddtrace/testing/internal/pytest/_discovery.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+from ddtrace.internal.settings import env
+from ddtrace.testing.internal.pytest.utils import _get_test_parameters_json
+from ddtrace.testing.internal.pytest.utils import item_to_test_ref
+from ddtrace.testing.internal.utils import asbool
+
+
+log = logging.getLogger(__name__)
+
+_ENV_ENABLED = "DD_TEST_OPTIMIZATION_DISCOVERY_ENABLED"
+_ENV_OUTPUT_PATH = "DD_TEST_OPTIMIZATION_DISCOVERY_FILE"
```
Comment (mabdinur) [inline ddtrace/testing/internal/pytest/_discovery.py]: Should we define these configs in `ddtrace.internal.settings`? It would be nice if we could centalize the loading of configs in one place

File: ddtrace/testing/internal/pytest/_discovery.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+
+@pytest.hookimpl(tryfirst=True)  # type: ignore[misc]
+def pytest_collection_finish(session: pytest.Session) -> None:
+    if not is_discovery_mode_enabled():
+        return
+
+    workspace_path: t.Optional[Path] = None
+    try:
+        from ddtrace.testing.internal.git import get_workspace_path
```
Comment (mabdinur) [inline ddtrace/testing/internal/pytest/_discovery.py]: Can we avoid inlining this import. This is a sign of a circular dependency. It would be nice if we could detangle this

File: tests/testing/internal/pytest/test_pytest_discovery.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+import json
+from pathlib import Path
+import sys
+
+from _pytest.pytester import Pytester
+import pytest
+
+
+class TestDiscoveryMode:
```
Comment (mabdinur) [inline tests/testing/internal/pytest/test_pytest_discovery.py:13]: Can we add a test case for when no items are collected (I believe the output path should be empty)?

File: ddtrace/testing/internal/pytest/_discovery.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        if condition:
+            return True
+    return False
+
+
+def _get_suite_source_file(item: pytest.Item, workspace_path: t.Optional[Path]) -> str:
+    item_path = Path(item.path if hasattr(item, "path") else getattr(item, "fspath", "")).absolute()
+    if workspace_path is not None:
+        try:
+            # TODO: use item_path.relative_to(workspace_path).as_posix() on Windows
```
Comment (mabdinur) [inline ddtrace/testing/internal/pytest/_discovery.py]: Can we address this todo, this could be a bug in windows

File: ddtrace/testing/internal/pytest/_discovery.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        from ddtrace.testing.internal.git import get_workspace_path
+
+        workspace_path = get_workspace_path()
+    except Exception:
+        log.debug("Could not determine workspace path for test discovery", exc_info=True)
+
+    output_path = _get_output_path()
+    output_path.parent.mkdir(parents=True, exist_ok=True)
+
+    with open(output_path, "w") as f:
```
Comment (mabdinur) [inline ddtrace/testing/internal/pytest/_discovery.py]: Should we open this file with an encoding?

File: ddtrace/testing/internal/pytest/plugin.py [hunk 6]
```diff
@@ -27,11 +25,14 @@
 from ddtrace.testing.internal.logging import catch_and_log_exceptions
 from ddtrace.testing.internal.logging import setup_logging
 from ddtrace.testing.internal.offline_mode import get_offline_mode
+from ddtrace.testing.internal.pytest._discovery import is_discovery_mode_enabled
 from ddtrace.testing.internal.pytest.bdd import BddTestOptPlugin
 from ddtrace.testing.internal.pytest.benchmark import BenchmarkData
 from ddtrace.testing.internal.pytest.benchmark import get_benchmark_tags_and_metrics
 from ddtrace.testing.internal.pytest.hookspecs import TestOptHooks
 from ddtrace.testing.internal.pytest.report_links import print_test_report_links
+from ddtrace.testing.internal.pytest.utils import _encode_test_parameter  # noqa: F401
```
Comment (mabdinur) [inline ddtrace/testing/internal/pytest/plugin.py]: nit: can we remove this import. it isn't used

**Review by mabdinur — APPROVED** [review]

**Review by anmarchenko — APPROVED** [review]
